### PR TITLE
Add tool for generating example MVP files from a template

### LINKED ISSUE #37730: Create python script to create template MVP files
**Description**
We want to create a python script which can be run from the command line to generate a couple of template files that can be used for an Model-View-Presenter widget. The script will be used for training purposes, and can also be used in future by developers wanting to create a widget using the MVP pattern. A link to the script should be put in the MVP documentation.

**Acceptance Criteria**
- A python script which generates a Model, View and Presenter files which contains the best MVP practises (interface classes, passing view and model to presenter as constructor arguments, no Qt in the presenter)
- The python script should have a few arguments:
      - The name to use for the classes
      - The location to output the files to
      - Whether to generate python files, or cpp files for MVP
- The script should be documented and a link provided to it

### PR DESCRIPTION
**Pull request by robertapplin**
### Description of work
This PR adds a cmd line tool that can be used to generate MVP files from several templates in python and C++.

Fixes #37730

### To test:
Follow the new instructions in the MVPDesign file to create an example python MVP widget, and an example C++ MVP widget.

You should be able to generate the Model View Presenter files Ok, and then launch the example widget using the instructions.

*This does not require release notes* because **it is a tool for developers**

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: dev-docs/source/MVPDesign.rst [hunk 1]
```diff
@@ -179,6 +179,68 @@ in C++, or `unittest.mock
...
+`ExampleView.cpp`, `ExampleView.h`, `ExamplePresenter.cpp` and `ExamplePresenter.h`.
+An additional file named `main.cpp` and a `CMakeLists.txt` will be generated. These
+files can be used to build the widget as follows:
+
+.. code-block:: sh
+
+  mkdir buildmvp
+  cd buildmvp
+  cmake ..
+  cmake --buildmvp .
```
Comment (robertapplin) [inline dev-docs/source/MVPDesign.rst:230]:
```suggestion
  cmake --build .
```

File: dev-docs/source/MVPDesign.rst [hunk 2]
```diff
@@ -179,6 +179,68 @@ in C++, or `unittest.mock
...
+  cd buildmvp
+  cmake ..
+  cmake --build .
+
+The example widget can then be opened with:
+
+.. code-block:: sh
+
+  cd buildmvp
+  ./Debug/launch.exe
```
Comment (adriazalvarez) [inline dev-docs/source/MVPDesign.rst:237]: On windows it generates by default with visual studio. But in MacOs (and I guess Linux too) it uses Makefiles. 
In this case, the output does not have .exe extension and is not in debug folder. So could you add a small line indicating that in Unix systems the launch instructions for cpp should be:
```suggestion
  ./launch
```

Also the `./Debug/launch.exe` won't work in windows unless called from a shell or bash. You can tip to use the GitBash to run this example on windows.

File: dev-docs/source/MVPDesign.rst [hunk 3]
```diff
@@ -179,6 +179,68 @@ in C++, or `unittest.mock
 can set expectations in the unit tests for certain methods to be
 called, and with certain arguments.
 
+MVP Template tool
+#################
+
+The `template.py script <https://github.com/mantidproject/mantid/blob/main/tools/MVP/template.py>`__
```
Comment (adriazalvarez) [inline dev-docs/source/MVPDesign.rst:185]: Is probably obvious, but these instructions should be run under mantid-developer environment. is it worth adding this comment somewhere?

File: dev-docs/source/MVPDesign.rst [hunk 4]
```diff
@@ -179,6 +179,68 @@ in C++, or `unittest.mock
...
+The example widget can then be opened with:
+
+.. code-block:: sh
+
+  cd buildmvp
+  ./Debug/launch.exe
+
+The `main.cpp` and a `CMakeLists.txt` files are intended as an example for how you can
+build, and then instantiate your widget. If you are refactoring or creating a new
+widget for Mantid, the headers and cpp files should be included in the relevant
```
Comment (adriazalvarez) [inline dev-docs/source/MVPDesign.rst:241]: Not sure how this tool is generally going to be used, but in the case that it is used for development, it is not necessary to generate a CMakeLists.txt file, is it worth adding an argument to template.py for omitting it?

File: tools/MVP/templates/View.cpp.in [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  auto layout = new QVBoxLayout();
+  auto button = new QPushButton("Increment", this);
+  m_label = new QLabel("0", this);
+
+  layout->addWidget(button);
+  layout->addWidget(m_label);
+
+  setLayout(layout);
+
+  connect(button, SIGNAL(clicked()), this, SLOT(notifyButtonClicked()));
```
Comment (adriazalvarez) [inline tools/MVP/templates/View.cpp.in:29]: I know that it is not used much at the moment in Mantid, but should we try to enforce the usage of new signal/slot syntax in this example case?
https://wiki.qt.io/New_Signal_Slot_Syntax

**Review by adriazalvarez — CHANGES_REQUESTED** [review]
I'm giving just some initial feedback, as others are going to check this too. 
I have run the examples on Windows and MacOS. It runs without any problem. And the example GUIs are also running fine. It's a very nice and simple way to kickstart the work on MVP pattern, thank you for creating this. 
The code to generate the files is clear to follow, I haven't found any problem.  
I have added some minor suggestions, or in some cases queries, that I have found when using the classes. 
The other suggestions is that I think the example files are missing a `.ui` file. For complicated interfaces we don't generate the ui elements programmatically, I think it would be a nice addition to add an example of how to use ui files to work with the views. But I understand you want to keep it simple. But then I don't know if these files are for easing development or for illustrating the pattern (in which case is fine as it is).

File: dev-docs/source/MVPDesign.rst (hunk 4 above)
Comment (robertapplin) [reply dev-docs/source/MVPDesign.rst:241]: Thats a good idea! The setup files will not be needed in a lot of cases. I've added an `--include-setup` option

**Comment by robertapplin** [comment]
I'll get back to you on the `.ui` file suggestion, I think that could be useful!

**Comment by sf1919** [comment]
I've given this a 6.12 milestone as it's not a priority to get this in by end of Thursday (12th September)

**Comment by robertapplin** [comment]
> * Apart from that, I think there are potential improvements that could be made to the script in the future. For instance, implementing a way to automatically add getter/setters to the view, and creating event handling for every event in the view, each with a corresponding handler in the presenter.

Great, thank you for testing and reading through the documentation! 

I agree this sounds like it could be a useful addition in the future. I think I will leave it for a future PR as I am unsure I will have enough time to look into this before the MVP training in October, but it sounds like something we could explore further!

**Comment by robertapplin** [comment]
> * Although this may be common in the current code, I think the presenter raw pointer in the view needs to be deleted manually as it is not a QObject. The QObject model automatically handles the lifecycle of QObjects; if you add a child to a parent, the parent is responsible for deleting it. However, I have observed that the presenter in C++ could be a non-QObject, which could lead to a memory leak. I am unsure how critical this is, as it has been implemented like that across the project's interfaces.

This is something of a subtlety in the code. When we subscribe the presenter to the view, we are passing the view a non-owning raw pointer to the presenter (meaning that the presenter is owned by something else and will therefore be deleted when necessary). If we were to attempt to manually delete the presenter in the view destructor, we would get a segmentation fault because we would be deleting the presenter twice (once in the view destructor, and a second time when the unique ptr holding the presenter goes out of scope). It is the unique ptr holding onto the presenter which has ownership of the object

**Review by adriazalvarez — APPROVED** [review]
Thanks for the changes. It's looking good to me .

**Pull request merged**